src/flashbar/collapsible-flashbar.tsx

```diff
@@ -0,0 +1,392 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import clsx from 'clsx';
+import React, { useCallback, useEffect, useLayoutEffect, useRef, useState } from 'react';
+import customCssProps from '../internal/generated/custom-css-properties';
+import { Flash } from './flash';
+import { FlashbarProps, FlashType, StackedFlashbarProps } from './interfaces';
+import InternalIcon from '../icon/internal';
+import { TransitionGroup } from 'react-transition-group';
+import { Transition } from '../internal/components/transition';
+import useFocusVisible from '../internal/hooks/focus-visible';
+import { getVisualContextClassname } from '../internal/components/visual-context';
+
+import styles from './styles.css.js';
+import { getFlashTypeCount, getItemSelection, StackableItem } from './utils';
+import { animate, getDOMRects } from '../internal/animate';
+import { useUniqueId } from '../internal/hooks/use-unique-id';
+import { IconProps } from '../icon/interfaces';
+import { sendToggleMetric } from './internal/analytics';
+import { componentName, useFlashbar } from './common';
+import { warnOnce } from '../internal/logging';
+import LiveRegion from '../internal/components/live-region';
+
+export { FlashbarProps };
+
+// If the number of items is equal or less than this value,
+// the toggle element will not be displayed and the Flashbar will look like a regular single-item Flashbar.
+const maxNonCollapsibleItems = 1;
+
+export default function CollapsibleFlashbar({ items, ...restProps }: FlashbarProps & StackedFlashbarProps) {
+  const [enteringItems, setEnteringItems] = useState<ReadonlyArray<FlashbarProps.MessageDefinition>>([]);
+  const [exitingItems, setExitingItems] = useState<ReadonlyArray<FlashbarProps.MessageDefinition>>([]);
+  const [isFlashbarStackExpanded, setIsFlashbarStackExpanded] = useState(false);
+
+  const getElementsToAnimate = useCallback(() => {
+    const flashElements = isFlashbarStackExpanded ? expandedItemRefs.current : collapsedItemRefs.current;
+    return { ...flashElements, toggleButton: toggleElementRef.current };
+  }, [isFlashbarStackExpanded]);
```

Comment:
This was hinted at me by my IDE. The reasoning is that there are memoized functions that used to depend on `elementsToAnimate` which would be recreated on every render. And the reason why I use `useCallback` and not `useMemo` (I tried this too) is because it would depend on the arrays `expandedItemRefs.current` and `collapsedItemRefs.current`, which don't trigger updates when they mutate. The latter is solved by invoking the function every time I need it.

So it all has to do with the subtleties of object identity JS: if not memoizing at all, I would get trouble from generating too many objects; if memoizing the values, I would not update every time I should; only memoizing as a function does the trick. I don't think it's a big overhead —it should be an unproblematic function and it will be recreated only when the expand/collapse state is toggled.